server/src/main/java/org/apache/iotdb/db/query/executor/groupby/SlidingWindowGroupByExecutorFactory.java

```diff
@@ -0,0 +1,116 @@
+      String aggrFuncName, TSDataType dataType, boolean ascending) {
+    if (aggrFuncName == null) {
+      throw new IllegalArgumentException("AggregateFunction Name must not be null");
+    }
+
+    switch (aggrFuncName.toLowerCase()) {
+      case SQLConstant.SUM:
+      case SQLConstant.AVG:
+      case SQLConstant.COUNT:
+        return new SmoothQueueSlidingWindowGroupByExecutor(dataType, aggrFuncName, ascending);
+      case SQLConstant.MAX_VALUE:
+        return new MonotonicQueueSlidingWindowGroupByExecutor(
+            dataType, aggrFuncName, ascending, (o1, o2) -> ((Comparable<Object>) o1).compareTo(o2));
+      case SQLConstant.MIN_VALUE:
+        return new MonotonicQueueSlidingWindowGroupByExecutor(
+            dataType, aggrFuncName, ascending, (o1, o2) -> ((Comparable<Object>) o2).compareTo(o1));
+      case SQLConstant.EXTREME:
+        return new MonotonicQueueSlidingWindowGroupByExecutor(
+            dataType,
+            aggrFuncName,
+            ascending,
+            (o1, o2) -> {
+              TSDataType resultDataType = getResultDataType(dataType, aggrFuncName);
+              Comparable<Object> extVal = (Comparable<Object>) o1;
+              Comparable<Object> absExtVal = (Comparable<Object>) getAbsValue(o1, resultDataType);
+              Comparable<Object> candidateResult = (Comparable<Object>) o2;
+              Comparable<Object> absCandidateResult =
+                  (Comparable<Object>) getAbsValue(o2, resultDataType);
+              if (absExtVal.compareTo(absCandidateResult) == 0) {
+                return extVal.compareTo(candidateResult);
+              } else {
+                return absExtVal.compareTo(absCandidateResult);
+              }
+            });
+      case SQLConstant.MIN_TIME:
+      case SQLConstant.FIRST_VALUE:
+        return !ascending
+            ? new EmptyQueueSlidingWindowGroupByExecutor(dataType, aggrFuncName, ascending)
+            : new NormalQueueSlidingWindowGroupByExecutor(dataType, aggrFuncName, ascending);
+      case SQLConstant.MAX_TIME:
+      case SQLConstant.LAST_VALUE:
+        return !ascending
+            ? new NormalQueueSlidingWindowGroupByExecutor(dataType, aggrFuncName, ascending)
+            : new EmptyQueueSlidingWindowGroupByExecutor(dataType, aggrFuncName, ascending);
+      default:
+        throw new IllegalArgumentException("Invalid Aggregation Type: " + aggrFuncName);
+    }
+  }
+
+  private static Object getAbsValue(Object v, TSDataType resultDataType) {
+    switch (resultDataType) {
+      case DOUBLE:
+        return Math.abs((Double) v);
+      case FLOAT:
+        return Math.abs((Float) v);
+      case INT32:
+        return Math.abs((Integer) v);
+      case INT64:
+        return Math.abs((Long) v);
+      default:
+        throw new UnSupportedDataTypeException(java.lang.String.valueOf(resultDataType));
+    }
+  }
+
+  private static TSDataType getResultDataType(TSDataType dataType, String aggrFuncName) {
```

Comment:
There's a same method in `TypeInferenceUtils.getAggrDataType`. I think we can remove this and leave that.

Reply:
fixed